pkg/controller/attach_detach.go

```diff
@@ -554,42 +499,51 @@ func (r *ReconcileAttachDetach) recreateAzVolumeAttachment(ctx context.Context,
+					azVolumeAttachment = &diskv1beta1.AzVolumeAttachment{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: azVolumeAttachmentName,
+							Labels: map[string]string{
+								consts.NodeNameLabel:   nodeName,
+								consts.VolumeNameLabel: *volumeName,
+							},
+							Finalizers: []string{consts.AzVolumeAttachmentFinalizer},
+						},
+						Spec: diskv1beta1.AzVolumeAttachmentSpec{
+
+							VolumeName:    *volumeName,
+							VolumeID:      pv.Spec.CSI.VolumeHandle,
+							NodeName:      nodeName,
+							RequestedRole: diskv1beta1.PrimaryRole,
+							VolumeContext: map[string]string{},
 						},
-						Finalizers: []string{consts.AzVolumeAttachmentFinalizer},
-					},
-					Spec: diskv1beta1.AzVolumeAttachmentSpec{
-
-						VolumeName:    *volumeName,
-						VolumeID:      pv.Spec.CSI.VolumeHandle,
-						NodeName:      nodeName,
-						RequestedRole: diskv1beta1.PrimaryRole,
-						VolumeContext: map[string]string{},
-					},
-					Status: diskv1beta1.AzVolumeAttachmentStatus{
-						State: azureutils.GetAzVolumeAttachmentState(volumeAttachment.Status),
-					},
-				}
-				if azVolumeAttachment.Status.State == diskv1beta1.Attached {
-					azVolumeAttachment.Status.Detail = &diskv1beta1.AzVolumeAttachmentStatusDetail{
-						Role: diskv1beta1.PrimaryRole,
 					}
-				}
-				_, err := r.controllerSharedState.azClient.DiskV1beta1().AzVolumeAttachments(r.controllerSharedState.objectNamespace).Create(ctx, &azVolumeAttachment, metav1.CreateOptions{})
-				if err != nil {
-					klog.Errorf("failed to create AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *volumeName, nodeName, err)
+					azVolumeAttachment, err = r.controllerSharedState.azClient.DiskV1beta1().AzVolumeAttachments(r.controllerSharedState.objectNamespace).Create(ctx, azVolumeAttachment, metav1.CreateOptions{})
+					if err != nil {
+						klog.Errorf("failed to create AzVolumeAttachment (%s) for volume (%s) and node (%s): %v", azVolumeAttachmentName, *volumeName, nodeName, err)
+						return syncedVolumeAttachments, volumesToSync, err
+					}
+				} else {
+					klog.Errorf("failed to get AzVolumeAttachment (%s): %v", azVolumeAttachmentName, err)
 					return syncedVolumeAttachments, volumesToSync, err
 				}
-			} else {
-				klog.Errorf("failed to get AzVolumeAttachment (%s): %v", azVolumeAttachmentName, err)
+			}
+
+			azVolumeAttachment.Status = diskv1beta1.AzVolumeAttachmentStatus{
+				State: azureutils.GetAzVolumeAttachmentState(volumeAttachment.Status),
+			}
+			if azVolumeAttachment.Status.State == diskv1beta1.Attached {
+				azVolumeAttachment.Status.Detail = &diskv1beta1.AzVolumeAttachmentStatusDetail{
+					Role: diskv1beta1.PrimaryRole,
+				}
+			}
+			// update status
+			_, err = r.controllerSharedState.azClient.DiskV1beta1().AzVolumeAttachments(r.controllerSharedState.objectNamespace).UpdateStatus(ctx, azVolumeAttachment, metav1.UpdateOptions{})
```

Comment:
Now that create and status update are two separate operations, the controller may see the created object *before* the status update occurs. In this case, it is ok because this method is only called during recovery and the reconciler will ignore the creation event. However, are there other cases where this would be problematic?